src/compiler/crystal/syntax/parser.cr

```diff
@@ -3702,14 +3704,19 @@ module Crystal
 
     def parse_visibility_modifier(modifier)
       doc = @token.doc
-
       next_token_skip_space
-      exp = parse_op_assign
 
-      modifier = VisibilityModifier.new(modifier, exp)
-      modifier.doc = doc
-      exp.doc = doc
-      modifier
+      if @token.type == :NEWLINE
```

Comment:
1. Maybe `VisibilityModifier.new(modifier, nil)`. That, of course, will "break" some code, but then you have: if the `exp` is `nil` it's just `private` (or `protected`), if it's not `nil` then it's applying to some other node.
2. You'll have to store the current visibility in the TypeVisitor. Start with `nil`. Then when you encounter a `VisibilityModifier` with a `nil` exp, you set the current visibility to that one. Subsequent defs or declarations (like calls, that expand to macros) should check the current visibility and apply it.